[chewjh1234] iP

chewjh1234 opened the pull request:
# Tired
### _The Best Chatbot Ever_

> "This has the best chatbot ever"  -- Jing Heng, creator of Tired

Tired frees your mind of having to remember things you need to do, which means you will become less tired! It's,

* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use


All you need to do is,
1. download it from [here](https://github.com/chewjh1234/ip/releases/tag/A-Jar)
2. double click it
3. add your tasks
4. and let it manage them for you! :smirk: 

And it is **FREE** with **ABSOLUTELY NO ADS**

Features:
 - [ ] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

If you Java programmer, you can use it to practice Java too. Here's the `main` method: 

```java
public static void main(String[] args) {
        String fileName = "duke.txt";
        new Tired(fileName).run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

chewjh1234 committed:
Add Level-0

chewjh1234 committed:
Add Level-1

chewjh1234 committed:
Add Level-2

chewjh1234 committed:
Add Level-3

chewjh1234 committed:
Add Level-4

chewjh1234 committed:
Add Level-5

chewjh1234 committed:
Add Level-6, enum and ArrayList

chewjh1234 committed:
Add Level-6, fix bugs

chewjh1234 committed:
Fix bugs for 'delete'

chewjh1234 committed:
Add A-TextUiTesting

chewjh1234 committed:
Add Level-7

chewjh1234 committed:
Add Level-7. Remove Duke.java.

chewjh1234 committed:
Merge branch 'branch-Level-7'

chewjh1234 committed:
Add Level-8

chewjh1234 committed:
Merge branch 'branch-Level-8'

chewjh1234 committed:
Add A-MoreOOP

chewjh1234 committed:
Add A-Packages

chewjh1234 committed:
Merge commit 'ed6d4d25bd4e34461919ffe785a6f10f4c087552'

chewjh1234 committed:
Add Gradle

chewjh1234 committed:
Merge branch 'branch-A-Gradle'

chewjh1234 committed:
Add A-JUnit and test cases

chewjh1234 committed:
Modify build.gradle

chewjh1234 committed:
Add A-JavaDoc

chewjh1234 committed:
Add A-CodingStandard

chewjh1234 committed:
Add Level-9

chewjh1234 committed:
Merge branch 'branch-A-JavaDoc'

chewjh1234 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/ToDo.java
#	src/main/java/duke/Ui.java

chewjh1234 committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/Parser.java
#	src/main/java/duke/Ui.java

chewjh1234 committed:
Fix accidentally-pushed merge conflicts.

chewjh1234 committed:
Add JavaDoc for find

coderhuang559 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,224 @@
+        }
+
+        String taskType = parts[0].trim();
+        String done = parts[1].trim();
```
`done` is ambiguous. Perhaps `taskStatus` conveys the String contents more clearly?

rngzhi replied:
```suggestion
String taskStatus = parts[1].trim();
``

coderhuang559 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,224 @@
+     * @return The Task object created from the input.
+     * @throws DukeException If there is an error in parsing the input or creating the task.
+     */
+    public static Task parseCreateTaskInput(String input) throws DukeException {
```
Maybe something along the lines of `parseStringToTask` is clearer. Found the method name a little confusing.

rngzhi replied:
```suggestion
public static Task parseStringToTask(String input) throws DukeException {
```

coderhuang559 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,102 @@
+public class Task {
+    // inspired by partial solution on website
+    protected String description;
+    protected boolean isDone;
```
Boolean name sounds like boolean name. Great!

coderhuang559 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,102 @@
+     *
+     * @return true if the task is done, false if it's not done.
+     */
+    public boolean isDone() { return this.isDone;
```
Do take note of Egyptian brackets

coderhuang559 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,102 @@
+    /**
+     * Marks the task as undone.
+     */
+    public void markAsUndone() {
```
Not really an issue, but consider naming as `UnmarkAsDone` instead, to match with naming in the rest of the project.

coderhuang559 commented on src/test/java/duke/ParserTest.java:
```diff
@@ -0,0 +1,124 @@
+public class ParserTest {
+
+    @Test
+    public void testIsCreateTaskCommandTodo() {
```
Maybe you can name your test methods according to conventions as mentioned in the notes. Will be clearer and easier to understand the purpose of each test.

rngzhi replied:
Perhaps the test method name should follow the format: "...Test", where "..." represent the original method name.

coderhuang559 reviewed (commented):
Overall good job! A few aspects can be improved.

rngzhi commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,49 @@
+     */
+    public Event(String description, String start, String end) throws DukeException {
+        super(description, TaskType.EVENT);
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
```
Should constant "formatter" be extracted out as a static variable?

rngzhi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,224 @@
+            throw new DukeException("I'm sorry, but I don't know what that means :-(");
+        }
+        return true;
+    }
```
Perhaps a more precise method name here? Given that this method not only checks if the command is valid, but also performs the corresponding operation. Or perhaps extract the execution part and put in another new method.

HollaG replied:
I agree with the above

rngzhi commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,224 @@
+            String end = parts[2].substring(3).trim();
+            newTask = new Event(taskDetails, start, end);
+            break;
+        }
```
Perhaps a default block makes it more general?

rngzhi commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,98 @@
+                Files.createFile(dataPath);
+            }
+        } catch (IOException e) {
+            System.err.println(e.getMessage());
```
Perhaps error messages should be parsed to UI?

rngzhi commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,102 @@
+package duke;
```
Perhaps a package called duke.task can be created?

rngzhi commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,103 @@
+        tasks.get(taskIndex).markAsUndone();
+        System.out.println("OK, I've marked this task as not done yet:");
+        System.out.println(tasks.get(taskIndex).toString());
+        System.out.println("Now you have " + getSize() + " task(s) in the list.");
```
Same issue as the error message one.

rngzhi commented on src/main/java/duke/Tired.java:
```diff
@@ -0,0 +1,43 @@
+package duke;
+
+enum TaskType {
+    TODO, DEADLINE, EVENT
+}
```
I like how you make use of enum type.

rngzhi reviewed (commented):
Overall, you do a good job. The code is well structured though some variable and method names could be improved.

HollaG commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,49 @@
+     * @throws DukeException If the description, start time, or end time is empty.
+     */
+    public Event(String description, String start, String end) throws DukeException {
+        super(description, TaskType.EVENT);
```
should description be .trim() here as well?
```suggestion
        super(description.trim(), TaskType.EVENT);
```

HollaG commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,224 @@
+            if (taskIndex >= 0 && taskIndex < tasks.getSize()) {
+                tasks.unmarkTask(taskIndex);
+            }
+            return true;
```
instead of using `substring()` which might not be immediately obvious what it's doing / cause bugs if the command changes, consider using `string.replace("find ", "")`

HollaG commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,224 @@
+                break;
+            }
+            if (done.equals("0") && task != null) {
+                task.isDone = false;
```
try not to use public property access here. instead, possibly use a `setIsDone()` or `setIsUndone()` method in `Task`

HollaG replied:
edit: you already have a `markAsDone` method you can use

HollaG commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,224 @@
+
+        if (taskDetails.startsWith("todo")) {
+            taskType = TaskType.TODO;
+            taskDetails = taskDetails.substring(4).trim();
```
Similar to my comment above about using `.replace("todo ", "")` instead of substring

HollaG commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,102 @@
+     * @return A formatted string representing the task in the format:
+     *         "taskType || doneStatus || taskDescription || start || end"
+     */
+    public String taskToFileString() {
```
Perhaps a more general name could be used, e.g. if you'd want to store the task as a string in something that's not a file.
```suggestion
    public String encodeTaskAsString() {
```

HollaG commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,102 @@
+            end = ((Event) this).end.format(DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm"));
+        }
+
+        return taskType + " || " + doneStatus + " || " + taskDescription + " || " + start + " || " + end;
```
Can actually abstract out the " || " into a static string called DIVIDER or something similar so if you change the divider in one place you don't have to change it everywhere else

HollaG commented on src/main/java/duke/Tired.java:
```diff
@@ -0,0 +1,43 @@
+    TODO, DEADLINE, EVENT
+}
+
+public class Tired {
```
Missing Javadocs :(

HollaG reviewed (commented):
LGTM!

Overall good! Minor coding style errors (indentation, naming) etc as noted by myself and others.

chewjh1234 committed:
Fix code style issues stated in PR comments

chewjh1234 committed:
Add JavaFX GUI

chewjh1234 committed:
Merge branch 'branch-Level-10'

chewjh1234 committed:
Add A-CheckStyle

chewjh1234 committed:
Merge branch 'branch-A-CheckStyle'

chewjh1234 committed:
Add Assertions to various classes

There are several important assumptions made.
For example, indexes of the tasks should not be greater than the size of
the task list.

If these assumptions fail, the program will stop running unexpectedly.

Let's add assertions to prove that the assumptions are valid and prevent
unexpected errors from happening.

chewjh1234 committed:
Improve code quality by refactoring Parser, Ui and other classes

There are several methods that exceed 30 lines of code. Moreover,
there are multiple layers of abstraction within one method. This makes
it very hard to read.

Refactoring code by extracting the various logical components into other
methods helps with code readability.

Let's improve the code readability by extracting logical components in
these classes.

chewjh1234 committed:
Merge pull request #1 from chewjh1234/branch-A-Assertions

Add Assertions to various classes

chewjh1234 committed:
Merge branch 'master' of https://github.com/chewjh1234/ip

chewjh1234 committed:
Merge branch 'master' into branch-A-CodeQuality

chewjh1234 committed:
Merge pull request #2 from chewjh1234/branch-A-CodeQuality

Improve code quality by refactoring Parser, Ui and other classes

chewjh1234 committed:
Merge branch 'master' of https://github.com/chewjh1234/ip

chewjh1234 committed:
Add ability to archive task list

Users are able to store the task list on their computers.

However, they are not able to start the application on a clean slate
with an empty task list unless they delete everything in the task list.

Let's add an archive file so that users can choose to store their
task lists temporarily and use the application without past tasks.

Improvements to this feature can be made in the future.

chewjh1234 committed:
Minor bug fixes

Fix task list not appearing fully.

chewjh1234 committed:
Add Ui.png into docs

chewjh1234 committed:
Update README.md

chewjh1234 committed:
Update README.md

chewjh1234 committed:
Update README.md

chewjh1234 committed:
Fix bugs
